src/template-expr.adoc

```diff
@@ -0,0 +1,255 @@
+
+
+==== Other Scalars are Literals
+
+When a template is a non-symbol Ion scalar, it denotes a literal value, and the template expands
+into that value. Any annotations on the template are included in the output.
+
+
+==== Lists and Structs are Quasi-Literals
+
+When a template is an Ion list or struct, it denotes a quasi-literal of the same type.
+We say “quasi” literal because the elements of the container are treated as templates, not
+literal values.
+
+When a template is a list, it expands into a list with the same annotations.
+The elements of the list-template are each treated as templates themselves.
+Each sub-template may produce any number of values, and the resulting streams are all
+concatenated to produce the output list.
+
+----
+[1, [2, 3], 4]       ⇒ [1, [2, 3], 4]
+[1, (values 2 3), 4] ⇒ [1, 2, 3, 4]
+[1, (values), 3]     ⇒ [1, 3]
+----
+
+When a template is a struct, it expands into a struct with the same annotations.
+The struct-template’s field names are treated as literals, and field values are treated as
+sub-templates, and the output struct contains the given names and their associated sub-template
+expansions.
+
+Field-value sub-templates MAY produce multiple values.
+When a sub-template produces more than one result, then the output struct will have more than one
+field with the same name.
+When a sub-template produces no results, then nothing is added to the output.
+
+----
+{a:(values 1 2)} ⇒ {a:1, a:2}   // or, equivalently, {a:2, a:1}
+{f:(values)}     ⇒ {}
+----
+
+
+==== S-expressions are Operator Invocations
+
+The template language uses S-expressions to denote operations using Lisp-style prefix notation.
+The first element of the S-expression must be a symbol that identifies the operator, and the
+meaning of subsequent elements depends on the operator.
+
+Operators come in two varieties: special forms and macro invocations.
+
+
+=== Special Forms
+
+Special forms are operators that cannot be expressed as macros, because some parts of their
+syntax are not recursively-expanded templates, as all macro arguments are.
+
+We use `*bold monospace*` when naming these special forms, to distinguish them from macro names.
+
+In the descriptions below, `_template_` subforms accept any template-language form.
+In all such cases, sub-templates are expanded only when indicated.
+
+
+==== Preventing Evaluation
+
+===== `*literal*`
+
+[{nrm}]
+----
+(*literal* _datum_)
+----
+
+Produces _datum_ as-is, preventing the operand from being evaluated as a template.
+
+For example, `(*literal* [1, (values 2 3), 4])` produces `[1, (values 2 3), 4]`; both the list
+and the S-expression are treated as literal, constant data, not as template expressions to be
+expanded.
+
+
+==== Conditionals
+
+These special forms allow output to vary based on whether a template produces zero, one, or more
+values.
+
+
+===== `*if_void*`
+
+[{nrm}]
+----
+(*if_void* _template1_ _template2_ _template3_)
```

Comment:
The current names are ok, but I wonder if we could use something more descriptive.

```suggestion
(*if_void* _target_template_ _true_template_ _false_template_)
```

Reply:
I've updated this to use descriptive subscripts.